### ISSUE TITLE
On Windows, pressing Ctrl-C during an Android `build` or `run` causes briefcase to hang

### ISSUE THREAD

**Issue by mhsmith** [beeware/briefcase#809]
Since the recent Briefcase logging changes, I can no longer interrupt an Android `build` or `run` on Windows. @rmartin16: You might be interested in this.

The `build` command launches two java.exe processes: one Gradle client and one Gradle daemon. Pressing Ctrl-C causes the daemon process to exit, but there is no feedback in the console, and the "Building" waitbar continues to animate. Pressing Ctrl-C a second time displays the message `Terminate batch job (Y/N)?`, but entering `y` has no effect. Killing the other java.exe process (the Gradle client) has no effect either. The only way I can get back to the command line prompt is to kill both briefcase.exe and python.exe. (This happens in both the cmd and MSYS2 terminals.)

There's a similar problem when `briefcase run` is displaying the Logcat. Ctrl-C kills the adb.exe process, but there is no feedback in the console, and Briefcase continues to run until I kill it with something outside of the terminal. (This only happens for me in a cmd terminal: MSYS2 works fine.)

**Expected behavior**
In all situations, Ctrl-C should promptly shut everything down and return to the command line prompt.

**Environment:**
 - Operating System: Windows 10
 - Python version: 3.8
 - Software versions:
   - Briefcase: 8df2d112e

**Comment by rmartin16** [comment]
hmm.....`cmd.exe` strikes again. The `Terminate batch job (Y/N)?` prompt apparently shows up because all these Android tools are being invoked via `bat` files and interpretation is being deferred to `cmd.exe`. There are third party tools to deal with this prompt but msft doesn't seem to offer a straightforward way to disable it. Making matters worse....the `KeyboardInterrupt` is sent to Python whether you answer Yes or No....

That said, the problem isn't nearly as severe for me........until I tried Python 3.8. On Python 3.10, pressing CTRL+C more times or even pressing "Y" and ENTER during the Wait Bar got me back to the prompt; exiting `logcat` with CTRL+C works no problem.

There must be something fundamentally different about Python 3.8 (likely in subprocess I imagine). Especially if CTRL+C doesn't work for exiting during `logcat`....I'm gonna dig some more.

**Comment by rmartin16** [comment]
This is a bit unexpected tbh. The hang is coming from joining the output streamer thread; the same thread implementation is used for the console printing during the Wait Bar and printing `adb logcat` output.

Potentially interesting:
 - Even if you remove the `join`, the Python interpreter just hangs after `briefcase` exits.
 - If you kill the `adb logcat` process outside of Python, `briefcase` (and Python) exit normally.
 - If the output streamer thread is started as a daemon thread, briefcase exits normally on CTRL+C.

Still trying to identify why Python is seemingly holding on to this thread...

(I'm largely ignoring the terminate batch job prompt problem right now since this hang seems a bit more significant.)

**Comment by rmartin16** [comment]
- https://github.com/python/cpython/issues/89437
- https://bugs.python.org/issue45274
- https://bugs.python.org/issue21822

I think that's it; victor et al reduced the likelihood of this occurring back to 3.9.

When you watch the threads for `python.exe` (in something like SysInternals procexp) after the CTRL-C is sent, you can clearly see the output streamer thread disappearing while Python's main thread hangs.

Setting `daemon=True` and removing the `join` (or adding a `timeout`) in the `finally` clause of `Subprocess.stream_output` avoids this bug. I suppose these changes could technically be gated by python version...

As for the `Terminate batch job (Y/N)?` prompt, this [discussion](https://stackoverflow.com/a/50892407/8549606) seems like a decent summary of the situation. Given this user experience of CTRL+C is pretty bad, we should probably avoid wrapping `bat` files in Wait Bars.....or convince google to ship `ps1` scripts as well 🙃

**Comment by rmartin16** [comment]
Another thought came to mind....

So, the whole problem (IIUC) is receiving the CTRL+C while `join` is blocking; this situation introduces the locking race condition and Python thinks the thread continues existing forever.

However, if we replace the `join` in the `try` block in `stream_output` with a sleep loop and then do the `join` in the `KeyboardInterrupt` catch, we avoid the lock race condition and get largely the same outcome....(half baked maybe but I think plausible)

Something like this (although the two while loops should be combined):
```python
      try:
          output_streamer.start()
          if stop_func:
              while output_streamer.is_alive() and not stop_func():
                  time.sleep(0.1)
          else:
              while output_streamer.is_alive():
                  time.sleep(1)
      except KeyboardInterrupt:
          output_streamer.join()
      finally:
          self.cleanup(label, popen_process)
          output_streamer.join()
```

Also, to be clear....this is WIndows only issue.

**Comment by freakboy3742** [comment]
Thanks for that investigation @rmartin16.

I can't think of any reason why the streamer thread has any reason to persist beyond the life of the parent process; on that basis, making the streamer thread a daemon thread would seem to be the right approach regardless. Adding a timeout would seem like a good safety mechanism too. I'll defer to empirical evidence when it comes to evaluating whether the restructuring around `join()` that you've suggested is worthwhile.

Beyond that - my general philosophy is that when a platform makes something difficult to do, you should stop fighting the platform. That would seem to suggest disabling wait bars on Windows for those tools that are batch files would be worthwhile. Would it be plausible to stop/suspend the waitbar when invoking `subprocess.*` if the first arg ends in `.bat`?

**Comment by rmartin16** [comment]
Don't tell anyone I wrote this but.....

```python
        # HAAAACCKKKK
        if str(args[0]).endswith(".bat") and self.command.input._active_wait_bar:
            self.command.input._active_wait_bar.stop()
            self.command.input.is_output_controlled = False
```

Naively putting that at the top of `Subprocess.run()` and running `briefcase build android` does allow for a much more graceful handling of `cmd`'s terminate batch prompt.....so, I think it _may_ be plausible.

That said, we were depending on the Wait Bar's as a proxy mechanism to get spawned process output in to the log.....so, a lot in the air I guess.....this was a card close to the bottom of the house in a way...

**Comment by mhsmith** [comment]
Thanks for the investigation. If getting rid of the "Terminate batch job?" message means we lose the process output from the log file, then I think we'll just have to live with the message. As long as the user can enter Y and it actually terminates, it's not a major problem. Or [the discussion](https://stackoverflow.com/a/50892407/8549606) you linked says you can just press Ctrl-C a second time, which is even easier.

**Comment by rmartin16** [comment]
I think to help prevent conflating the two issues in play and help focus this issue, I've tried to separate them. There are also at least two separate fixes needed for what's been identified.

I separated out the related issue of `cmd`'s terminate batch prompt in to:
- #811

**Issue Summary**
As detailed in https://github.com/python/cpython/issues/89437, Python can deadlock on Windows while `join`ing a thread and receiving CTRL+C at the same time. That is why it becomes necessary to kill `python.exe` manually.

This deadlock can manifest in `briefcase` while the output streamer is running (such as while the Wait Bar is active); for a few different reasons, the output streamer runs in a thread. Therefore, `briefcase`'s main thread is blocked on a `join` for the streamer's thread when the CTRL+C is received and the issue occurs.

**Solutions**
Fortunately, the issue has a foreseeable sunset since it was significantly mitigated in Python 3.9+. So, a solution should perhaps only apply to Python 3.7,3.8 and be removed when support for those python versions is dropped from `briefcase`.

1) Daemon thread
Running the output streamer as a daemon thread avoids the deadlock since Python will then skip trying to `join` the thread and will simply allow the OS to kill it when the `python.exe` main thread exits. This also means `Subprocess.stream_output()` can no longer `join` the thread itself.
A primary consideration for the `join` (at least in the `finally` block) was to ensure all spawned process output had been printed to the console by the streamer thread before allowing `briefcase` to move forward potentially resulting in interspersed output.

2) Replace the `join` in the `try` block
Replace the `join` with some form of sleeping loop waiting for `thread.is_alive()` to return `False`. Initial experimentation showed differences in user experience with this method.....however, such a sleeping loop is already in play if `stop_func` is used. Therefore, maybe it would be better to converge these methods of waiting on the output streamer and ensure both have the same semantics.
This is likely an existing issue but a major underlying assumption of output streaming is the CTRL+C to `briefcase` will be propagated to the process for which output is being streamed. This is usually true but it doesn't have to be such as when `start_new_session` is used. This is the situation for iOS app log streaming; however, since it also uses `stop_func`, it avoids `join`ing the output streamer. At any rate, maybe this all means moving away from `join` is a better solution for this use-case in general.

**Comment by freakboy3742** [comment]
(3) sounds like it might be a good idea generally. Less moving parts means less things likely to break. There's no need to introduce additional threading complexity unless it's actually needed (as it is with the iOS case). I'm not especially concerned if we lose a little buffered output in the case of a CTRL-C - if you're relying on the exact timing of CTRL-C to control log output, you possibly need to reconsider what you're doing.

Option (2) also looks like it might be a reasonable approach. At the very least, it would mean we maintain parity between the two implementations. That said, if we adopt (3), I'm guessing the need to do (2) will disappear.

**Comment by mhsmith** [comment]
I think the "converge" option (2) is the best choice, because two implementations  would complicate an already-complex area. And if one implementation is only used on a particular platform, then it'll be too easy for developers working on the other platforms to break it without noticing. Even if the breakage gets caught by the CI tests, it'll still waste time compared to discovering it at an earlier stage.

> Initial experimentation showed differences in user experience with this method

In what way?

I think we should also make the streamer a daemon thread, so that no matter what happens there's no chance of the user ending up with an uninterruptible Briefcase process. For example, I think this could happen in the current version if the subprocess ignores `popen_process.terminate`, and then the user causes a second KeyboardInterrupt while `cleanup` is blocked in `popen_process.wait`.

But to prevent interspersed output, `stream_output` should still give the thread a chance to finish. You said that adding a `timeout` argument to `join` prevented the deadlock, but that wasn't the experience of [this person](https://bugs.python.org/issue21822). So for safety it looks like we should avoid `join` altogether.

How about this?
```python
        try:
            output_streamer.start()
            if not stop_func:
                stop_func = lambda: False
            while output_streamer.is_alive() and not stop_func():
                time.sleep(0.1)
        except KeyboardInterrupt:
            pass  # allow CTRL+C to gracefully stop streaming
        finally:
            self.cleanup(label, popen_process)
            deadline = time.time() + TIMEOUT  # `cleanup` should use the same timeout.
            while (time.time() < deadline) and output_streamer.is_alive():
                time.sleep(0.1)
```

**Comment by mhsmith** [comment]
@rmartin16: Are you planning to make a PR for this in the next few days? If not, please let me know and I'll do it myself, because we'd like to make a new Briefcase release soon and it needs to include a fix for this. Thanks.

**Comment by rmartin16** [comment]
I agree converging the implementations is probably best. Setting `daemon=true` since the thread can safely be killed whenever is fine along with the benefits you mention.

>>    Initial experimentation showed differences in user experience with this method

> In what way?

May have wrote this too hastily since "user experience" may not be the best descriptor. Nonetheless, the `adb` process to capture the logs was seemingly being killed by Python instead of the CTRL+C with these changes (likely from `cleanup`)....so, Its exit semantics were different and `briefcaese` was registering that as some sort of failure.

> Are you planning to make a PR for this in the next few days?

I'll take a crack at it today given this feedback.
